MM-825 Replace FullName field with separate FirstName and LastName fields and repurpose the existing FullName as Nickname

hmhealey opened the pull request:
Changes made:
1) Renamed all existing references to the FullName field to Nickname
2) Added FirstName and LastName fields, populated from the previous FullName if possible, and use them wherever it makes sense
3) Added GetDisplayName helper function to be more consistent how we're referring to users in the UI. Currently, this is to use the nickname if possible and fall back to the username if that's not available.

Since the first two commits (for 1 above) are a bunch of find and replaces, it may be easier to review this on a commit by commit basis since they're fairly self-contained.

coreyhulen commented on model/user.go:
```diff
@@ -225,6 +236,14 @@ func (u *User) AddNotifyProp(key string, value string) {
+func (u *User) GetDisplayName() string {
+	if u.Nickname != "" {
+		return u.Nickname
+	} else {
```
should also have case for FullName?

hmhealey replied:
Oh yeah, I could see adding that. I wasn't thinking of it at the time since it was just Nickname and then username originally, but with the new First/LastName fields, it makes sense to include that as a second case.

coreyhulen commented on api/user.go:
```diff
@@ -207,7 +207,7 @@ func fireAndForgetWelcomeEmail(name, email, teamName, link string) {
 		subjectPage := NewServerTemplatePage("welcome_subject", link)
 		bodyPage := NewServerTemplatePage("welcome_body", link)
-		bodyPage.Props["FullName"] = name
+		bodyPage.Props["Nickname"] = name
```
I think you also need to modify the api/templates files

hmhealey replied:
Nothing in the template files reference FullName or any of variation of it so I don't believe those properties are currently used anyway.

hmhealey commented:
Updated the display name so that it can be a user's full name when their nickname hasn't been set. Rebased onto match the master branch to fix merge conflicts.

crspeller commented on web/react/stores/user_store.jsx:
```diff
@@ -177,21 +177,15 @@ var UserStore = assign({}, EventEmitter.prototype, {
-      if (user.notify_props.all === "true") keys.push('@all');
-      if (user.notify_props.channel === "true") keys.push('@channel');
+    if (user.notify_props && user.notify_props.mention_keys) keys = keys.concat(user.notify_props.mention_keys.split(','));
+    if (user.first_name && user.notify_props.first_name === "true") keys.push(user.first_name);
```
What about my nickname? That's the one I want to use.

hmhealey replied:
I can add a mention key for nickname, but I'd like to do that as a separate pull request since this one is pretty big already.

crspeller replied:
Sounds good.

crspeller commented:
Looks good. Just to clarify, why are we using Firstname Lastname Nickname instead of just Fullname and Nickname? It seems like Fullname would be better for localization because not all names fit into the firstname lastname constraint.

hmhealey commented:
We split first name and last name due to cases where we can't infer the first name properly like someone named Mary Ann Smith since before we were just taking the first word as the first name which is just Mary as opposed to Mary Ann in this example.

crspeller commented:
But why do we need to infer the first name? If we have display name (nickname), then we can just use that.

hmhealey commented:
There's a few cases where we're using the first name for things like emails "Bill has invited you to join" and for mentions.

crspeller commented:
Couldn't we change those instances to display name?

hmhealey commented:
That's a very good point. I've made a note so we can discuss this later today during meetings.

hmhealey commented:
Leaving this as first/last name since that's probably better for integrating with other systems. If we end up needing just a single full name field, it's a lot easier to go back to that than going from full name to first/last name anyway.

hmhealey commented:
Rebased to fix any conflicts and ready to merge
